File: composeApp/src/commonMain/kotlin/org/sesacteamproject/passmate/ui/mypage/ReputationViewModel.kt
package org.sesacteamproject.passmate.ui.mypage

import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.sesacteamproject.passmate.auth.domain.usecase.IsSignedInUseCase
import org.sesacteamproject.passmate.core.model.AppResult
import org.sesacteamproject.passmate.mvi.MviViewModel
import org.sesacteamproject.passmate.user.domain.usecase.GetMyBadgesUseCase
import org.sesacteamproject.passmate.user.domain.usecase.GetMyGradeUseCase
import org.sesacteamproject.passmate.user.domain.usecase.GetMyProfileUseCase

class ReputationViewModel(
    private val getMyGradeUseCase: GetMyGradeUseCase,
    private val getMyBadgesUseCase: GetMyBadgesUseCase,
    private val getMyProfileUseCase: GetMyProfileUseCase,
    private val isSignedInUseCase: IsSignedInUseCase
) : MviViewModel<ReputationUiState, ReputationAction, ReputationEvent>(ReputationUiState()) {

    private var hasEntered = false

    private fun onEnter() {
        if (hasEntered) {
            return
        }
        hasEntered = true
        // 회원 전용 가드 — 서버 검증이 최종 권위지만 UX상 진입 시 먼저 로그인 유도한다 (규칙 §8)
        if (!isSignedInUseCase.invoke()) {
            viewModelScope.launch {
                _event.emit(ReputationEvent.RequireSignIn)
            }
        } else {
            load()
        }
    }

    private fun load() {
        _uiState.update { it.copy(isLoading = true, loadFailed = false) }
        viewModelScope.launch {
            val gradeDeferred = async { getMyGradeUseCase.invoke() }
            val badgesDeferred = async { getMyBadgesUseCase.invoke() }
            val profileDeferred = async { getMyProfileUseCase.invoke() }
            val gradeResult = gradeDeferred.await()
            val badgesResult = badgesDeferred.await()
            val profileResult = profileDeferred.await()

            if (gradeResult is AppResult.Success && badgesResult is AppResult.Success && profileResult is AppResult.Success) {
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        loadFailed = false,
                        profile = profileResult.value,
                        grade = gradeResult.value,
                        badges = badgesResult.value
                    )
                }
            } else {
                _uiState.update { it.copy(isLoading = false, loadFailed = true) }
            }
        }
    }

    override fun onAction(action: ReputationAction) {
        when (action) {
            is ReputationAction.Enter -> onEnter()
            is ReputationAction.Retry -> load()
        }
    }
}

File: iosApp/iosApp/ui/mypage/ReputationViewModel.swift
import Combine
import Foundation
import Shared

// Compose ReputationViewModel.kt 미러 — 내 명성·뱃지 로드 (M-09)
final class ReputationViewModel: ObservableObject {
    private let getMyGradeUseCase: GetMyGradeUseCase

    private let getMyBadgesUseCase: GetMyBadgesUseCase

    private let isSignedInUseCase: IsSignedInUseCase

    @Published private(set) var uiState = ReputationUiState()

    let event = PassthroughSubject<ReputationEvent, Never>()

    private var hasEntered = false

    private func onEnter() {
        if hasEntered {
            return
        }
        hasEntered = true
        // 회원 전용 가드 — 서버 검증이 최종 권위지만 UX상 진입 시 먼저 로그인 유도 (규칙 §8)
        if isSignedInUseCase.invoke() {
            load()
        } else {
            event.send(.requireSignIn)
        }
    }

    private func load() {
        uiState.isLoading = true
        uiState.loadFailed = false
        getMyGradeUseCase.invoke { [weak self] gradeResult, gradeError in
            DispatchQueue.main.async {
                guard let self else { return }
                let grade = (gradeResult as? AppResultSuccess<AnyObject>)?.value as? MyGrade

                if gradeError == nil, let grade {
                    self.loadBadges(grade: grade)
                } else {
                    self.uiState.isLoading = false
                    self.uiState.loadFailed = true
                }
            }
        }
    }

    private func loadBadges(grade: MyGrade) {
        getMyBadgesUseCase.invoke { [weak self] badgesResult, badgesError in
            DispatchQueue.main.async {
                guard let self else { return }
                let badges = (badgesResult as? AppResultSuccess<AnyObject>)?.value as? [Badge]

                if badgesError == nil, let badges {
                    self.uiState.isLoading = false
                    self.uiState.loadFailed = false
                    self.uiState.grade = grade
                    self.uiState.badges = badges
                } else {
                    self.uiState.isLoading = false
                    self.uiState.loadFailed = true
                }
            }
        }
    }

    func action(_ action: ReputationAction) {
        switch action {
        case .enter:
            onEnter()
        case .retry:
            load()
        }
    }

    init(
        getMyGradeUseCase: GetMyGradeUseCase,
        getMyBadgesUseCase: GetMyBadgesUseCase,
        isSignedInUseCase: IsSignedInUseCase
    ) {
        self.getMyGradeUseCase = getMyGradeUseCase
        self.getMyBadgesUseCase = getMyBadgesUseCase
        self.isSignedInUseCase = isSignedInUseCase
    }
}
